### PR TITLE
Create operator-tools module to get token reader-secret

### PR DESCRIPTION
**Pull request by shanchunyang0919**
Regression/smoke tests pass

### CODE REVIEW

**Review by Laci21 — APPROVED** [review]
lgtm (took a quick glance only)

**Review by nishantapatil3 — APPROVED** [review]
lgtm!

File: pkg/secret/secretgetter.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+	Client                  client.Client
+	ServiceAccountName      string
+	ServiceAccountNamespace string
+}
+
+type readerSecretGetter struct {
+	options *ReaderSecretGetterOption
+}
+
+func NewReaderSecretGetter(opt *ReaderSecretGetterOption) (SecretGetter, error) {
```
Comment (waynz0r) [inline pkg/secret/secretgetter.go:50]:
```suggestion
func NewReaderSecretGetter(client client.Client) (SecretGetter, error) {
```

File: pkg/secret/secretgetter.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+	corev1 "k8s.io/api/core/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type SecretGetter interface {
+	Get() (*K8sSecret, error)
```
Comment (waynz0r) [inline pkg/secret/secretgetter.go:32]:
```suggestion
	Get(objectKey client.ObjectKey) (K8sSecret, error)
```

File: pkg/secret/secretgetter.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+type K8sSecret struct {
```
Comment (waynz0r) [inline pkg/secret/secretgetter.go:35]:
```suggestion
type K8sSecret interface {
        GetToken() []byte
        GetCACert() []byte
}

type k8sSecret struct {
```

File: pkg/secret/secretgetter.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+	}
+
+	if opt.ServiceAccountNamespace == "" || opt.ServiceAccountName == "" {
+		return nil, errors.New("service account name and namespace should be set for reader-secret getter")
+	}
+
+	return &readerSecretGetter{options: opt}, nil
+}
+
+func (r *readerSecretGetter) Get() (*K8sSecret, error) {
```
Comment (waynz0r) [inline pkg/secret/secretgetter.go:66]:
```suggestion
func (r *readerSecretGetter) Get(objectKey client.ObjectKey) (K8sSecret, error) {
```

File: pkg/secret/secretgetter.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+	readerSecret := &K8sSecret{
+		Token:  secretObj.Data["token"],
+		CACert: secretObj.Data["ca.crt"],
+	}
+
+	return readerSecret, nil
+}
+
+// backoff waiting of the K8s Secret object to be created
+var defaultBackoff = wait.Backoff{
```
Comment (waynz0r) [inline pkg/secret/secretgetter.go:144]: make this overridable

**Review by waynz0r — CHANGES_REQUESTED** [review]

**Comment by shanchunyang0919** [comment]
Thanks for reviewing @waynz0r. I have pushed some new fixes.

File: pkg/secret/secretgetter.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+	// After K8s v1.24, Secret objects containing ServiceAccount tokens are no longer auto-generated, so we will have to
+	// manually create Secret in order to get the token.
+	// Reference: https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.24.md#no-really-you-must-read-this-before-you-upgrade
+	return r.getOrCreateReaderSecretWithServiceAccount(ctx, sa)
+}
+
+func (r *readerSecretGetter) getReaderSecretServiceAccount(ctx context.Context, saObjectKey client.ObjectKey) (*corev1.ServiceAccount, error) {
+	sa := &corev1.ServiceAccount{}
+	saRef := types.NamespacedName{
```
Comment (waynz0r) [inline pkg/secret/secretgetter.go:83]: saObjectKey can be used directly as it is types.NamespacedName already

File: pkg/secret/secretgetter.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+		err = r.client.Create(ctx, secretObj)
+		if err != nil {
+			return nil, errors.WrapIfWithDetails(err, "creating kubernetes secret failed", "namespace",
+				secretObjRef.Namespace,
+				"secret name",
+				secretObjRef.Name)
+		}
+
+		// wait for token-controller to create token for the reader secret
+		backoffWaitSecret := wait.Backoff{
```
Comment (waynz0r) [inline pkg/secret/secretgetter.go:132]: still not overridable by the user, it can be moved to an overridable option using functional option pattern

**Review by waynz0r — CHANGES_REQUESTED** [review]

**Review by waynz0r — APPROVED** [review]

**Comment by shanchunyang0919** [comment]
Thanks for approval and reviews @waynz0r @Laci21 @nishantapatil3

**Pull request merged**